diff -ruN NetworkManager-0.6.5_p20070823.orig/configure.in NetworkManager-0.6.5_p20070823/configure.in
--- NetworkManager-0.6.5_p20070823.orig/configure.in	2008-01-09 21:20:03.000000000 -0600
+++ NetworkManager-0.6.5_p20070823/configure.in	2008-01-09 22:10:28.000000000 -0600
@@ -144,6 +144,11 @@
 
 PKG_CHECK_MODULES(HAL, hal >= 0.5.0)
 
+PKG_CHECK_EXISTS([hal >= 0.5.10],[have_hal_0_5_10=yes],[have_hal_0_5_10=no])
+if test "$have_hal_0_5_10" = "yes"; then
+	AC_DEFINE([HAVE_HAL_0_5_10],[1],[Define if we have hal 0.5.10 or newer])
+fi
+
 if test x"$with_gnome" != xno; then
 	PKG_CHECK_MODULES(GTK, gtk+-2.0)
 
diff -ruN NetworkManager-0.6.5_p20070823.orig/src/NetworkManager.c NetworkManager-0.6.5_p20070823/src/NetworkManager.c
--- NetworkManager-0.6.5_p20070823.orig/src/NetworkManager.c	2008-01-09 21:20:03.000000000 -0600
+++ NetworkManager-0.6.5_p20070823/src/NetworkManager.c	2008-01-09 22:09:21.000000000 -0600
@@ -339,7 +339,11 @@
 {
 	DBusError		err;
 	DBusMessage *	reply = NULL;
-	guint32			status;
+#ifdef  HAVE_HAL_0_5_10
+	gint32		status = 1;
+#else
+	guint32		status = 1;
+#endif
 
 	g_return_if_fail (pcall != NULL);
 	g_return_if_fail (data != NULL);
@@ -357,10 +361,20 @@
 		dbus_error_free (&err);
 		goto out;
 	}
-
-	if (!dbus_message_get_args (reply, &err, DBUS_TYPE_UINT32, &status, DBUS_TYPE_INVALID)) {
+	/* Handle both HAL <=0.5.9 which uses UINT and HAL >= 0.5.10 which
+	 * uses INT.
+	 */
+#ifdef	HAVE_HAL_0_5_10
+	if (!dbus_message_get_args (reply, &err, DBUS_TYPE_INT32, &status, DBUS_TYPE_INVALID)) {
+#else
+	if (!dbus_message_get_args (repy, &err, DBUS_TYPE_UINT32, &status, DBUS_TYPE_INVALID)) {
+#endif
+	if (!ks_err_message || strcmp (ks_err_message, err.message)) {
 		nm_info ("Error getting killswitch power arguments: %s - %s", err.name, err.message);
-		dbus_error_free (&err);
+		g_free (ks_err_message);
+		ks_err_message = g_strdup (err.message);
+	}
+	dbus_error_free (&err);
 		goto out;
 	}
 
