diff -uNr libzvt-2.0.1.orig/libzvt/libzvt-2.0.pc.in libzvt-2.0.1/libzvt/libzvt-2.0.pc.in
--- libzvt-2.0.1.orig/libzvt/libzvt-2.0.pc.in	2006-04-30 20:46:59.000000000 +0200
+++ libzvt-2.0.1/libzvt/libzvt-2.0.pc.in	2006-04-30 20:47:19.000000000 +0200
@@ -7,5 +7,5 @@
 Description: libzvt
 Requires: gtk+-2.0 libart-2.0
 Version: @VERSION@
-Libs: -L${libdir} -lzvt-2.0 @UTIL_LIBS@
+Libs: -L${libdir} -lzvt-2.0
 Cflags: -I${includedir}/libzvt-2.0
diff -uNr libzvt-2.0.1.orig/libzvt/Makefile.in libzvt-2.0.1/libzvt/Makefile.in
--- libzvt-2.0.1.orig/libzvt/Makefile.in	2006-04-30 20:46:59.000000000 +0200
+++ libzvt-2.0.1/libzvt/Makefile.in	2006-04-30 20:48:30.000000000 +0200
@@ -112,7 +112,7 @@
 libzvt_2_0_la_SOURCES =  	libzvt.h 	libzvt-init.c 	gnome-login-support.h		gnome-login-support.c		lists.h				lists.c				subshell.h			subshell.c			subshell-includes.h		update.c			vt.c				vt.h				vtx.h				background.c			background.h			zvtterm.c			zvt-marshal.c			zvt-marshal.h			zvt-accessible.c		zvt-accessible.h		zvt-accessible-factory.c		zvt-accessible-factory.h
 
 
-libzvt_2_0_la_LIBADD = $(ZVT_LIBS)
+libzvt_2_0_la_LIBADD = $(ZVT_LIBS) $(UTIL_LIBS)
 
 libzvt_2_0_la_LDFLAGS =  	-version-info $(LT_CURRENT):$(LT_REVISION):$(LT_AGE)
 
