diff -pur dbus-1.0.2/dbus/dbus-spawn.c dbus-1.0.2.new/dbus/dbus-spawn.c
--- dbus-1.0.2/dbus/dbus-spawn.c	2006-12-11 21:21:09.000000000 +0200
+++ dbus-1.0.2.new/dbus/dbus-spawn.c	2007-10-29 08:31:56.256664752 +0200
@@ -551,9 +551,20 @@ babysitter_iteration (DBusBabysitter *si
     {
       int ret;
 
-      ret = _dbus_poll (fds, i, 0);
+      do
+        {
+          ret = _dbus_poll (fds, i, 0);
+        }
+      while (ret < 0 && errno == EINTR);
+
       if (ret == 0 && block)
-        ret = _dbus_poll (fds, i, -1);
+        {
+          do
+            {
+              ret = _dbus_poll (fds, i, -1);
+            }
+          while (ret < 0 && errno == EINTR);
+        }
       
       if (ret > 0)
         {
@@ -989,8 +1000,12 @@ babysit (pid_t grandchild_pid,
       pfds[1].fd = sigchld_pipe[READ_END];
       pfds[1].events = _DBUS_POLLIN;
       pfds[1].revents = 0;
-      
-      _dbus_poll (pfds, _DBUS_N_ELEMENTS (pfds), -1);
+
+      if (_dbus_poll (pfds, _DBUS_N_ELEMENTS (pfds), -1) < 0 && errno != EINTR)
+        {
+          _dbus_warn ("_dbus_poll() error: %s\n", strerror (errno));
+          exit (1);
+        }
 
       if (pfds[0].revents != 0)
         {
