diff -pur dbus-1.0.2/dbus/dbus-misc.c dbus-1.0.2.new/dbus/dbus-misc.c
--- dbus-1.0.2/dbus/dbus-misc.c	2006-12-11 21:21:05.000000000 +0200
+++ dbus-1.0.2.new/dbus/dbus-misc.c	2007-10-26 10:48:25.740378320 +0300
@@ -76,12 +76,14 @@ dbus_get_local_machine_id (void)
   char *s;
 
   s = NULL;
-  _dbus_string_init (&uuid);
+  if (!_dbus_string_init (&uuid))
+    return NULL;
+
   if (!_dbus_get_local_machine_uuid_encoded (&uuid) ||
       !_dbus_string_steal_data (&uuid, &s))
     {
       _dbus_string_free (&uuid);
-      return FALSE;
+      return NULL;
     }
   else
     {
