diff -pur dbus-1.0.2/dbus/dbus-connection.c dbus-1.0.2.new/dbus/dbus-connection.c
--- dbus-1.0.2/dbus/dbus-connection.c	2006-12-11 21:21:08.000000000 +0200
+++ dbus-1.0.2.new/dbus/dbus-connection.c	2007-10-25 16:59:56.912931368 +0300
@@ -1374,13 +1374,13 @@ _dbus_connection_unref_unlocked (DBusCon
 static dbus_uint32_t
 _dbus_connection_get_next_client_serial (DBusConnection *connection)
 {
-  int serial;
+  dbus_uint32_t serial;
 
   serial = connection->client_serial++;
 
-  if (connection->client_serial < 0)
+  if (connection->client_serial == 0)
     connection->client_serial = 1;
-  
+
   return serial;
 }
 
