diff -pur dbus-1.0.2/dbus/dbus-connection.c dbus-1.0.2.new/dbus/dbus-connection.c
--- dbus-1.0.2/dbus/dbus-connection.c	2006-12-11 21:21:08.000000000 +0200
+++ dbus-1.0.2.new/dbus/dbus-connection.c	2007-10-04 12:35:11.375634880 +0300
@@ -3057,9 +3057,6 @@ dbus_connection_send_with_reply (DBusCon
    if (!_dbus_connection_get_is_connected_unlocked (connection))
     {
       CONNECTION_UNLOCK (connection);
-
-      *pending_return = NULL;
-
       return TRUE;
     }
 
