diff -NurdB mergeant-0.64/src/query-editor.c mergeant-0.64-patched/src/query-editor.c
--- mergeant-0.64/src/query-editor.c	2006-10-12 03:11:16.000000000 -0500
+++ mergeant-0.64-patched/src/query-editor.c	2006-11-17 13:02:59.000000000 -0500
@@ -341,7 +341,7 @@
 		/* the query did not have any graph: create one */
 		graph = GDA_GRAPH (gda_graph_query_new (qedit->priv->query));
 		gda_graph_query_sync_targets (GDA_GRAPH_QUERY (graph));
-		gda_dict_assume_object (gda_object_get_dict (GDA_OBJECT (qedit->priv->query)), (GdaObject *) graph);
+		gda_dict_assume_object_as (gda_object_get_dict (GDA_OBJECT (qedit->priv->query)), (GdaObject *) graph, GDA_TYPE_GRAPH);
 		g_object_unref (graph);
 	}
 	wid = gnome_db_canvas_query_struct_new (qedit->priv->query, graph);
