diff -ruN pidgin-2.0.0.orig/libpurple/idle.c pidgin-2.0.0/libpurple/idle.c
--- pidgin-2.0.0.orig/libpurple/idle.c	2007-05-14 15:17:38.000000000 -0500
+++ pidgin-2.0.0/libpurple/idle.c	2007-05-14 23:24:50.000000000 -0500
@@ -92,6 +92,7 @@
 	purple_presence_set_idle(presence, FALSE, 0);
 }
 
+static gint time_until_next_idle_event;
 /*
  * This function should be called when you think your idle state
  * may have changed.  Maybe you're over the 10-minute mark and
@@ -111,13 +112,15 @@
  * 3. Report your current idle time to the IM server.
  */
 static gint
-check_idleness()
+check_idleness(void)
 {
 	time_t time_idle;
 	gboolean auto_away;
 	const gchar *idle_reporting;
 	gboolean report_idle;
 	GList *l;
+   gint away_seconds = 0;
+   static int no_away = 0;
 
 	purple_signal_emit(purple_blist_get_handle(), "update-idle");
 
@@ -153,16 +156,29 @@
 			time_idle = time(NULL) - last_active_time;
 	}
 
+   time_until_next_idle_event = IDLEMARK- time_idle; /* reasonable start upperbound */
+
+   if (auto_away || !no_away)
+      away_seconds = 60 * purple_prefs_get_int("/purple/away/mins_before_away");
+
 	if (auto_away &&
-		(time_idle > (60 * purple_prefs_get_int("/purple/away/mins_before_away"))))
+      (time_idle > away_seconds))
 	{
 		purple_savedstatus_set_idleaway(TRUE);
+      no_away = 0;
+      if (time_idle < away_seconds && (away_seconds - time_idle) < time_until_next_idle_event)
+         time_until_next_idle_event = away_seconds - time_idle;
 	}
-	else if (time_idle < 60 * purple_prefs_get_int("/purple/away/mins_before_away"))
+   else if (!no_away && time_idle < away_seconds)
 	{
 		purple_savedstatus_set_idleaway(FALSE);
+      no_away = 1;
+      if (time_idle < away_seconds && (away_seconds - time_idle) < time_until_next_idle_event)
+         time_until_next_idle_event = away_seconds - time_idle;
 	}
 
+
+
 	/* Idle reporting stuff */
 	if (report_idle && (time_idle >= IDLEMARK))
 	{
@@ -178,9 +194,21 @@
 			set_account_unidle(idled_accts->data);
 	}
 
+   if (time_until_next_idle_event < 0)
+      time_until_next_idle_event = IDLEMARK;
+
 	return TRUE;
 }
 
+/*
+ * Check idle and set the timer to fire at the next idle-worth event
+ */
+static gint
+check_idleness_time()
+{
+   idle_timer = purple_timeout_add(1000*(time_until_next_idle_event+1) , check_idleness_time, NULL);
+   return FALSE;
+}
 static void
 im_msg_sent_cb(PurpleAccount *account, const char *receiver,
 			   const char *message, void *data)
diff -ruN pidgin-2.0.0.orig/pidgin/gtkidle.c pidgin-2.0.0/pidgin/gtkidle.c
--- pidgin-2.0.0.orig/pidgin/gtkidle.c	2007-05-14 15:17:39.000000000 -0500
+++ pidgin-2.0.0/pidgin/gtkidle.c	2007-05-14 15:21:24.000000000 -0500
@@ -103,8 +103,12 @@
 
 	/* Query xscreensaver */
 	static XScreenSaverInfo *mit_info = NULL;
+   static int has_extension = -1;
 	int event_base, error_base;
 	if (XScreenSaverQueryExtension(GDK_DISPLAY(), &event_base, &error_base)) {
+      if (has_extension == -1)
+         has_extension = XScreenSaverQueryExtension(GDK_DISPLAY(), &event_base, &error_base);
+      if (has_extension) {
 		if (mit_info == NULL) {
 			mit_info = XScreenSaverAllocInfo();
 		}
