--- control-center/src/control-center.c	2006-05-21 19:54:09.000000000 -0600
+++ control-center/src/control-center.c.new	2006-08-16 19:07:29.000000000 -0600
@@ -106,7 +106,7 @@
         hidden = TRUE;
     }
 
-    AppShellData * app_data = appshelldata_new("preferences.menu", NULL, CONTROL_CENTER_PREFIX, GTK_ICON_SIZE_DIALOG);
+    AppShellData * app_data = appshelldata_new("settings.menu", NULL, CONTROL_CENTER_PREFIX, GTK_ICON_SIZE_DIALOG);
     generate_categories(app_data);
     GSList * actions = get_actions_list();
     layout_shell(app_data, _("Filter"), _("Groups"), _("Common Tasks"), actions, handle_static_action_clicked);
