--- configure	2007-03-14 16:39:54.000000000 +0300
+++ configure	2007-03-16 18:57:15.000000000 +0300
@@ -3635,21 +3635,21 @@
     ;;
 esac
 
-if test -n "0.23"; then
-    { echo "$as_me:$LINENO: checking for intltool >= 0.23" >&5
-echo $ECHO_N "checking for intltool >= 0.23... $ECHO_C" >&6; }
-
-    INTLTOOL_REQUIRED_VERSION_AS_INT=`echo 0.23 | awk -F. '{ print $ 1 * 1000 + $ 2 * 100 + $ 3; }'`
-    INTLTOOL_APPLIED_VERSION=`awk -F\" '/\\$VERSION / { print $ 2; }' ${ac_aux_dir}/intltool-update.in`
-    INTLTOOL_APPLIED_VERSION_AS_INT=`awk -F\" '/\\$VERSION / { split($ 2, VERSION, "."); print VERSION[1] * 1000 + VERSION[2] * 100 + VERSION[3];}' ${ac_aux_dir}/intltool-update.in`
-
-    { echo "$as_me:$LINENO: result: $INTLTOOL_APPLIED_VERSION found" >&5
-echo "${ECHO_T}$INTLTOOL_APPLIED_VERSION found" >&6; }
-    test "$INTLTOOL_APPLIED_VERSION_AS_INT" -ge "$INTLTOOL_REQUIRED_VERSION_AS_INT" ||
-	{ { echo "$as_me:$LINENO: error: Your intltool is too old.  You need intltool 0.23 or later." >&5
-echo "$as_me: error: Your intltool is too old.  You need intltool 0.23 or later." >&2;}
-   { (exit 1); exit 1; }; }
-fi
+#if test -n "0.23"; then
+#    { echo "$as_me:$LINENO: checking for intltool >= 0.23" >&5
+#echo $ECHO_N "checking for intltool >= 0.23... $ECHO_C" >&6; }
+#
+#    INTLTOOL_REQUIRED_VERSION_AS_INT=`echo 0.23 | awk -F. '{ print $ 1 * 1000 + $ 2 * 100 + $ 3; }'`
+#    INTLTOOL_APPLIED_VERSION=`awk -F\" '/\\$VERSION / { print $ 2; }' ${ac_aux_dir}/intltool-update.in`
+#    INTLTOOL_APPLIED_VERSION_AS_INT=`awk -F\" '/\\$VERSION / { split($ 2, VERSION, "."); print VERSION[1] * 1000 + VERSION[2] * 100 + VERSION[3];}' ${ac_aux_dir}/intltool-update.in`
+#
+#    { echo "$as_me:$LINENO: result: $INTLTOOL_APPLIED_VERSION found" >&5
+#echo "${ECHO_T}$INTLTOOL_APPLIED_VERSION found" >&6; }
+#    test "$INTLTOOL_APPLIED_VERSION_AS_INT" -ge "$INTLTOOL_REQUIRED_VERSION_AS_INT" ||
+#	{ { echo "$as_me:$LINENO: error: Your intltool is too old.  You need intltool 0.23 or later." >&5
+#echo "$as_me: error: Your intltool is too old.  You need intltool 0.23 or later." >&2;}
+#   { (exit 1); exit 1; }; }
+#fi
 
   INTLTOOL_DESKTOP_RULE='%.desktop:   %.desktop.in   $(INTLTOOL_MERGE) $(wildcard $(top_srcdir)/po/*.po) ; LC_ALL=C $(INTLTOOL_MERGE) -d -u -c $(top_builddir)/po/.intltool-merge-cache $(top_srcdir)/po $< $@'
 INTLTOOL_DIRECTORY_RULE='%.directory: %.directory.in $(INTLTOOL_MERGE) $(wildcard $(top_srcdir)/po/*.po) ; LC_ALL=C $(INTLTOOL_MERGE) -d -u -c $(top_builddir)/po/.intltool-merge-cache $(top_srcdir)/po $< $@'
